pyomo/contrib/edi/blackBoxFunctionModel.py

```diff
@@ -0,0 +1,1116 @@
+#  ___________________________________________________________________________
+#
+#  Pyomo: Python Optimization Modeling Objects
+#  Copyright (c) 2008-2023
+#  National Technology and Engineering Solutions of Sandia, LLC
+#  Under the terms of Contract DE-NA0003525 with National Technology and
+#  Engineering Solutions of Sandia, LLC, the U.S. Government retains certain
+#  rights in this software.
+#
+#  Development of this module was conducted as part of the Institute for
+#  the Design of Advanced Energy Systems (IDAES) with support through the
+#  Simulation-Based Engineering, Crosscutting Research Program within the
+#  U.S. Department of Energy’s Office of Fossil Energy and Carbon Management.
+#
+#  This software is distributed under the 3-clause BSD License.
+#  ___________________________________________________________________________
+
+import copy
+import pyomo
+import pyomo.environ as pyo
+from pyomo.environ import units as pyomo_units
+from pyomo.common.dependencies import attempt_import
+
+
+egb, egb_available = attempt_import(
+    "pyomo.contrib.pynumero.interfaces.external_grey_box"
+)
+
+from pyomo.common.dependencies import numpy, numpy_available
+from pyomo.common.dependencies import scipy, scipy_available
+
+
+if numpy_available:
+    import numpy as np
+else:
+    raise ImportError(
+        "pyomo.contrib.edi requires numpy to enable black box capability, fix with 'pip install numpy' "
+    )
+
+if scipy_available:
+    import scipy.sparse as sps
+else:
+    raise ImportError(
+        "pyomo.contrib.edi requires scipy to enable black box capability, fix with 'pip install scipy' "
+    )
+
+if egb_available:
+    from pyomo.contrib.pynumero.interfaces.external_grey_box import (
+        ExternalGreyBoxModel,
+        ExternalGreyBoxBlock,
+    )
+else:
+    raise ImportError(
+        "pyomo.contrib.edi requires pyomo.contrib.pynumero to be installed to enable black box capability, this should have installed with base pyomo"
+    )
+
+
+class BlackBoxFunctionModel_Variable(object):
```

Comment:
This class isn't described in the documentation

Reply:
This is not a user-facing class and therefore was not prioritized for the documentation.  It can be added in future developer documentation if desired